tests/suite/scripting/return.typ

```diff
@@ -85,3 +85,81 @@
   // Error: 16-16 expected semicolon or line break
   #return a + b Hello World
 ]
+
+--- return-discard-content ---
+// Test that discarding joined content is a warning.
+
+#let f() = {
+  state("hello").update("world")
+  // Warning: 3-16 unconditional explicit return value discards content
+  // Hint: 3-16 use `return` without a value to return the joined content
+  return "nope"
+}
+
+#test(f(), "nope")
+
+--- return-discard-content-nested ---
+
+#let f() = {
+  state("hello").update("world")
+  {
+    // Warning: 5-18 unconditional explicit return value discards content
+    // Hint: 5-18 use `return` without a value to return the joined content
+    return "nope"
+  }
+}
+
+#test(f(), "nope")
+
+--- return-discard-for ---
+
+#let f() = {
+  state("hello").update("world")
+  for x in range(3) {
+    return "nope"
+  }
+}
+
+#test(f(), "nope")
+
+--- return-discard-while ---
+
+#let f() = {
+  state("hello").update("world")
+  let i = 0
+  while i < 10 {
+    return "nope"
+  }
+}
+
+#test(f(), "nope")
+
+--- return-no-discard ---
+// Test that returning a joined value is not a warning.
+
+#let f() = {
+  state("hello").update("world")
+  return
+}
+
+#test(f(), state("hello").update("world"))
+
+--- return-discard-not-content ---
+// Test that non-joined value is not a warning.
```

Comment:
non-joined isn't really accurate, is it? since arrays can be joined?

Reply:
Oops, it should say `non-content joined value`